### ISSUE TITLE
[Suggestion] Additional assertions in Chialisp

### ISSUE THREAD

**Issue by SlugFiller** [Chia-Network/chia-blockchain#9796]
Since the addition of every assertion to Chialisp is a potential softfork (clients that do not recognize the new assertion can't sync), I want to make a list of assertions I feel are useful, and potentially necessary for certain types of contracts, while the network is still "new", and they can be added without much of a mess.

Alternately, if this sort of request is not appropriate, consider this a request for a review/triage process for new assertions.

I also considered making separate issues for each, but since each addition is a softfork, it's actually better to add assertions in bulk, rather than adding them one at a time. So I'm making the request in bulk as well.

The list is as follows:

ASSERT_COIN_HEIGHT (coin_id height) - Asserts the block height in which the coin `coin_id` was created. Makes ASSERT_HEIGHT_RELATIVE redundant, as it would be possible to combine ASSERT_COIN_HEIGHT, ASSERT_MY_COIN_ID and ASSERT_HEIGHT_ABSOLUTE to achieve the same (albeit, at increased cost, due to the extra inputs). While seemingly pointless on its own (since ASSERT_HEIGHT_RELATIVE exists), it is useful for the addition of any other height-based assertions, as it removes the need to have absolute and relative versions of each assertion. Having a `coin_id` parameter allows making assertions on parent and sibling coins as well (this could technically be done with announcements, but that requires the other coin to have a specific puzzle which announces height limitations, otherwise the announcement could simply be faked). An alternative is to make a special type of announcement which includes an unfakeable height assertion, e.g. CREATE_HEIGHT_ANNOUNCEMENT which creates an id of `sha256(height + coin_id + message)`.

ASSERT_MY_HEIGHT (height) - Same as above, but automatically inserting the current coin's id. Redundant, but can save a few bytes in the spend bundle due to not having to add the current coin's id as part of the solution.

ASSERT_MAX_HEIGHT_ABSOLUTE (height) - Asserts that a spend becomes invalid after `height` block has been reached. This allows having solution options that expire. The normal way to achieve this is to add a min-height solution that travels to a puzzle with the expiring solution disabled. But besides the necessary of having a constant monitor client to make the expiring travel, its also vulnerable to spend-race, if the expiring solution allows a spend-to-same-puzzle. It's possible to work around THAT by adding relative height limits to alternative solutions, but each workaround just adds more complexity and usage limitations, that could be easily solved by having a native max-height assertion. An example use would be if the singleton waiting room actively prevented reward claims after a certain height, thereby removing the claim/travel race without having to stop claimable payments to the singleton.

ASSERT_COIN_UNSPENT (coin_id height) - Asserts that at block height `height` coin `coin_id` already existed and was unspent. Currently, the only way to assert anything about another coin is to include it in the spend bundle, spending it in the process. However, when a coin has a permissive same-puzzle spend (used for assertions only), this can create spend-races. This is already visible in Singletons, where use of pay-to-singleton needs limitations to prevent spend-locking the singleton from the owner. While this is manageable with two users for the coin (e.g. Pool and farmer), if the user count increases, limitations do not scale particularly well. A way to passively use a coin as an information source without spending it is valuable. Additionally, setting a height to check means that even if there's a hyper-active writing spender (refreshing the coin on every block), it's still possible to assert data from the same block where the coin is spent, and even allow a few blocks safety margin, to allow the spend bundle to sit in the queue for a bit before being farmed. `height` is absolute, but can be made relative to the spend bundle by combining with ASSERT_MAX_HEIGHT_ABSOLUTE

ASSERT_COIN_SPENT (coin_id height) - Asserts that a coin exists, but was already spent before (or during) the block at height `height`.

ASSERT_COIN_NONEXISTENT (coin_id height) - Asserts that a coin did not yet exist at block height `height`.

Note: For the above 3 assertions, there are two possible ways to interpret the case where `height` is larger than the current block height. The first option is to interpret it as an implicit ASSERT_HEIGHT_ABSOLUTE. The other, is to clip it down to the height in which the spend bundle is farmed, thereby allowing to assert with height "infinity" (INT_MAX?) to imply "whenever this spend bundle is farmed". The latter option is more useful, because it's possible to explicitly add ASSERT_HEIGHT_ABSOLUTE if requiring the prior option.

ASSERT_BLOCK_HEADER_HASH (height hash) - Asserts that the block at height `height` has been farmed, and has header hash `hash`. Asserting this with a future block height allows adding a secure but deterministic random source, for use inside puzzles. The header hash is impossible to predict, since it depends on the winning farmers, but easy to verify once the block has been farmed. Having a source of randomness can be used to implement lottery coins, which choose their valid solutions based on the random seed provided by a future block's header hash.

**Comment by arvidn** [comment]
new conditions would be soft forks, meaning old clients *would* still be able to sync, they just wouldn't enforce the new conditions.
The way it works is that farmers won't include transactions with conditions or clvm operators they don't recognize in their mempool. i.e. when you farm a block, you sign off on knowing about all the conditions and operators that are used.

For purposes of validating blocks however, you accept that you may now know about all conditions and operators, and unknown ones are just treated as no-ops.

**Comment by SlugFiller** [comment]
@arvidn 
> For purposes of validating blocks however, you accept that you may now know about all conditions and operators, and unknown ones are just treated as no-ops.

This is making me wonder what would happen if a malicious farmer intentionally inserts an invalid condition. For instance, if a block contains a non-existing opcode, and a future update adds that opcode, and the assertion is invalid, does it trigger a rollback of the whole blockchain back to the block where the opcode was farmed? This doesn't help the farmer, but it's a distinct risk for the network.

**Comment by bramcohen** [comment]
Allowing assertions of max height would allow transactions to go from valid to invalid for a reason other than an input having been spent, which would have disastrous effects on the logic necessary to maintain a mempool. Having assertions of block ids would be even worse, because it would allow transactions to directly assert the state of other transactions which were none of their business.

**Comment by SlugFiller** [comment]
@bramcohen 
> Allowing assertions of max height would allow transactions to go from valid to invalid for a reason other than an input having been spent, which would have disastrous effects on the logic necessary to maintain a mempool.

I'd like some more information on what weird optimization is okay with a transaction being originally invalid (e.g. Due to bad puzzle solution), being valid and going invalid from a coin being spent (presumably, from a block being farmed), but would suddenly break down if a transaction is valid but goes invalid due to a block being farmed that does not specifically contains one of its coins.

Max height is a very powerful tool for contracts, and I can't see much of a difference between invalidating a bundle due to a block being added with a coin, vs invalidating a bundle due to a generic block addition.

If you run the puzzle before adding the bundle to the mempool, then the minimal max-height assertion is just one extra data point to add along with the input ids. If you run the puzzle after pulling from the mempool, then the max-height assertion is no harder to check than a coin announcement or signature assertion. Either way, I must be missing something, because I'm not seeing "disastrous effects"

> Having assertions of block ids would be even worse, because it would allow transactions to directly assert the state of other transactions which were none of their business.

Correct me if I'm wrong, but aren't the header hashes independent from the block content, and only dependent on the farmer key? From what I understood, there's a sort of independence that allows a farmer to rewrite a block already farmed, so long as another block wasn't farmed on top of it.

Regardless, I'm not sure what the issue with a transaction having "access" to information that is publicly available to any full node. If it's not secret from a node, why does it need to be a secret from a transaction? What's the attack vector, here?

Also, a transaction can't actually access a block of information, even if it had its hash, because it would require the solution to be as large as a block, rendering the transaction invalid.

**Comment by bramcohen** [comment]
Header hashes also depend on older blocks. It would be bad for there to be any way to pay for a reorg to happen.

The override principle is one of simplicity and predictability. If you really want to have any new conditions you should come up with a use case for one and make a compelling case for why it's necessary and useful. Without that we aren't even going to consider it.

**Comment by SlugFiller** [comment]
@bramcohen I can produce use cases. Let's start with the easy one. Use of ASSERT_COIN_UNSPENT and ASSERT_MAX_HEIGHT_ABSOLUTE for multi-user oracle. Here's how it can be used:

You make a CAT which TAIL received a puzzle hash from an oracle coin. The oracle is a singleton, which inner puzzle curries `(ORACLE_HASH pub_key value)`, and receives `(new_pub_key new_value)` as solution. It performs AGG_SIG_ME with `pub_key` of message `(new_pub_key new_value)`, and CREATE_COIN with the same puzzle but currying `(ORACLE_HASH new_pub_key new_value)`. In other words, a simple "Sign value on spend, and allow owner change" inner puzzle, which also contains the value in the puzzle hash.

The TAIL receives `(height pub_key puzzle amount parent_pub_key parent_value parent_amount parent_parent_id puzzle_solution)` as a solution (in addition to the regular TAIL inputs), and curries `(ORACLE_SINGLETON_STRUCT ORACLE_HASH)`. It calculates an oracle coin id as
```
(sha256
	(sha256
		parent_parent_id
		(puzzle-hash-of-curried-function (f ORACLE_SINGLETON_STRUCT)
			(puzzle-hash-of-curried-function ORACLE_HASH
				parent_value
				parent_pub_key
				ORACLE_HASH
			)
			(sha256tree ORACLE_SINGLETON_STRUCT)
		)
		parent_amount
	)
	(puzzle-hash-of-curried-function (f ORACLE_SINGLETON_STRUCT)
		(puzzle-hash-of-curried-function ORACLE_HASH
			(sha256tree puzzle)
			pub_key
			ORACLE_HASH
		)
		(sha256tree ORACLE_SINGLETON_STRUCT)
	)
	amount
)
```
Then it does ASSERT_UNSPENT using the `height` input, and ASSERT_MAX_HEIGHT_ABSOLUTE with `height+10`, and also runs `puzzle` with `puzzle_solution` and the regular TAIL inputs. The input `height` should be a block as recent as possible in which the oracle is unspent. This creates a spend bundle that is valid for 10 blocks, and uses a puzzle which hash is given by the oracle. If the oracle's value is changed via a transaction, the old value remains valid for 10 blocks, at which point a new spend bundle has to be created.

The alternative is to spend the oracle every time the TAIL is called. However, this means allowing the oracle to be no-op spent without any signature. That also means if two or more clients want to run the TAIL, they can't do it in the same block, or at the very least, they have to coordinate, otherwise they end up racing to be the first to spend the oracle coin. And if the oracle owner wants to modify it, they also have to race with any people wanting to run the TAIL. Imagine here a very permissive TAIL (e.g. send x*k mojo to puzzle_hash to mint k tokens) that could have hundreds of simultaneous users per block.

Another alternative is for the oracle owner to spend the oracle on every block, and for other clients to assert it from their own spend bundles without including it as an input themselves. However, this means the oracle owner has to keep no-op spending the oracle, regardless of whether or not it is used, as there's no way to predict if someone wants to use it in the next block. If the number of oracles running this way increases, each block suddenly becomes filled with oracle no-ops, and no actual transactions.

Without ASSERT_MAX_HEIGHT_ABSOLUTE, the oracle's value could never be "expired", since there's no way to check how many blocks old it is.

With the above solution, the oracle owner can change the value on any block, or even every block, and users can still use the puzzle in the value for the next 10 blocks. P.S. Since the value is a puzzle hash, presumably, the actual puzzle content is made public by an off-chain means.

For block id, as I've said, it's a good random source. It doesn't have to be a block id, it can be anything that's impossible to calculate faster than the network, e.g. challenge coming from timelords. Here's the use case:

You create a lottery coin. It curries a list of items of the form `(target_hash amount)`, and also two values `lock_height`, `unlock_height` where `unlock_height > lock_height`. Any person can spend it, passing as input an item of `(target_hash amount)`. It recreates itself with the item appended, and an amount equal the the sum of the amounts in all items. However, this form of spend also performs ASSERT_MAX_HEIGHT_ABSOLUTE with `lock_height`. After `lock_height` block has been farmed, it can no longer be appended to. A second form of spend receives `random_seed` and performs ASSERT_HEIGHT_ABSOLUTE on `unlock_height`, and ASSERT_BLOCK_HEADER_HASH with `unlock_height` and `random_seed`. Then, it uses `random_seed` as a random seed to pick one of the items from its list, weighted by their amount value. It then sends its full amount to the matching `target_hash` of the chosen item.

ASSERT_MAX_HEIGHT_ABSOLUTE ensures the lottery coin doesn't accept any more participants after `lock_height`. Without this, it would be possible to spend-lock the coin, preventing the reward from being collected, even after `unlock_height` is reached. A person could use this to continue spending 1 mojo, until the random value makes them the winner. ASSERT_BLOCK_HEADER_HASH is used to create a fair random, that any farmer can agree on once `unlock_height` is reached, but no one can predict that if they wage exactly a specific amount at `lock_height`, that they would necessarily win.

P.S. Sorry for a lot of hand-wavey. I could probably write all of this in the form of code, but aside from being time consuming, it would probably be actually harder to read. And may or may not have bugs.

**Comment by bramcohen** [comment]
Not sure what you're trying to do with TAILs there but it's a horrible abuse of their general purpose.

ASSERT_UNSPENT allows coins to dig into the state of coins which are none of their business and repeated oracle reports can be made by getting a singleton into a state where all it can do is repeatedly make the same announcement.

The lottery approach you're suggesting doesn't fix the most basic problem in lotteries which is that whoever puts in data last has a lot of control over the final output. Possible solutions include setting it up as a single elimination tournament in which participants do post-image commitment in each one-on-one competition (thus avoiding the very complicated denial of service problems with many participants where any one can refuse to show their preimage), using a VDF, or having a massive threshold signature. Using VDFs would require a new opcode of some kind.

The general kind of thing you want with ASSERT_MAX_HEIGHT_ABSOLUTE can be done by making a singleton have an anyone can spend with a min height thus assuring that most likely someone will spend it if there's enough attention.

**Comment by SlugFiller** [comment]
@bramcohen 
> Not sure what you're trying to do with TAILs there but it's a horrible abuse of their general purpose.
It's similar to the Delegated TAIL example, except the owner can "un-sign" a previously signed TAIL by spending the oracle, with a "waiting room" of 10 blocks.


> ASSERT_UNSPENT allows coins to dig into the state of coins which are none of their business and repeated oracle reports can be made by getting a singleton into a state where all it can do is repeatedly make the same announcement.
First, if ALL it can do is repeatedly make the same announcement, then it can't be edited. At a minimum, it needs to either make the same announcement, and allow the owner to edit the content.

Second, to make an announcement, it has to be spent. This means, if 2 users wants to use it in the same block, they can't do it with the same coin id, because it would be a double spend. One of them would have to detect that the oracle is spent once, calculate its next coin id, and use that as well. Only one of the users must do that, without coordinating, and their transaction now depends on the other user's transaction going through.

Now imagine it's 3 users. Now imagine it's 100 users, trying to use that same oracle, in the same block. Without coordinating, they must select an order between themselves, so they can each spend the same oracle with a different coin id. And they must rely on the farmer detecting that their spend bundles can be daisy-chained, and not drop any of their transaction from the mempool as having a non-existing input.

Alternately, each user can bombard the network with 10+ spend bundles, which only differ by the oracle's coin id. Since the other inputs can't double-spend, only one of the options would go through. But now the mempool is being flooded with 10+ transactions per user, of which only 1 per user is going to be valid. That's way closer to a "disastrous effect" on the mempool.

And all that without mentioning the owner's ability to edit. Although at least if it's just owner vs users, an ASSERT_HEIGHT_RELATIVE 2 can force them to take turns. The problem there is that it would now be completely impossible for 2 users to use an oracle in the same block. Every other block would basically be a race between 100 users trying to get their oracle-using spend bundle to go through.

The closest alternative solution I see is to allow a coin to be spent without being spent. i.e. add a DO_NOT_SPEND output that allows the coin to assert things only when solved, but without having to generate a new coin id, and without causing a double-spend if it is used multiple times in a spend bundle. Perhaps even the spend bundle can pre-emptively mark an input as non-spending. However, this seems to me like a massive change to the coin model that's way more impactful to the mempool/block building algorithm than simply allowing assertions on non-input coins.

> The lottery approach you're suggesting doesn't fix the most basic problem in lotteries which is that whoever puts in data last has a lot of control over the final output.

You can't call it "control" when neither the person doing it, nor any other participant, can predict if this last action improved or reduced their chance of winning, at least until several blocks AFTER that action has been taken (if even that). The last entrant makes no more difference to the final odds than the first one - "Removing" either changes the final selection, but in neither case can you calculate in advance who would benefit from that change.

> Possible solutions include setting it up as a single elimination tournament in which participants do post-image commitment in each one-on-one competition

There's no way to set it up in which "forfeiting" (not providing a base image) does not influence the result of a later match. Consider an entrant doing a Sybil attack by making multiple entries. In a "fair" system, the person's winning chances would be directly proportional to the number of entries they made (ergo, the amount they spent). However, suppose you have a tournament tree Alice-Bob1 - Charlie-Bob2 - Dan-Bob3 - Eric-Frank. After Alice, Charlie, Dan, Eric and Frank give their pre-images, Bob detects that the finals would be Alice vs Bob3 and Alice would win. However, if Bob2 forfeits to Charlie, then Alice would lose at Alice vs Charlie, and at Charlie vs Bob3, Bob3 would win. Bob has therefore found a way to increase his winning chances through "strategic forfeits". While there's still some luck involved (needing Charlie to be able to beat Alice, but not Bob3), Bob's odds are now slightly above the 3/8 he should have.

> using a VDF, or having a massive threshold signature. Using VDFs would require a new opcode of some kind.

Hmm... I assume you mean a VDF that uses the final contestant list as seed, thereby making it impossible for the final contestant to calculate whether they should participate or not, in the time it would take for the contest to close... I suppose that could work, depending on how expensive it is to verify a VDF. All contestants would have to calculate it to check if they won, as only the winning contestant would bother to provide the solution, and the puzzle itself has to verify the VDF, presumably without raking up a massive calculation cost in the process.

There's a few quirks, like the electricity cost to the contestant to calculate the VDF, and the fact that the VDF's difficulty has to be such that it can't be calculated within the entry time limit. But it works as a sort of "hard to predict random source".

Incidentally, I don't understand what you mean by "threshold signature", and how it provides a random source.

> The general kind of thing you want with ASSERT_MAX_HEIGHT_ABSOLUTE can be done by making a singleton have an anyone can spend with a min height thus assuring that most likely someone will spend it if there's enough attention.

This goes back to the ASSERT_UNSPENT situation: You still need to somehow assert the singleton still exists, you need to do it in a way that doesn't prevent it from doing the "disabling spend". If we're talking specifically about the "close lottery" case, then you can give the lottery a "closing" spend (removes the need for a singleton), but can face an issue of entrants doing a denial attack on the "closing" spend, extending the entry period.

If we're talking more generally, then you also have the fact that in a singleton mode, you can't base the checked height on an input parameter, and if it is used in a common, multi-user puzzle (e.g. a CAT's TAIL), then you end up with a user race again. i.e. You basically need a "height oracle" which suffers the exact same issues as the first example.

Even if some convoluted solution with only minor flaws exist, I don't see how reducing minor complexity in the assertions to increase massive complexity in the puzzles follows a principle of "simplicity and predictability". A max height check seems no different from an unspent input check. If the issue is that the max height is an output of the puzzle, this can be solved by forcing a bundle to declare its max height (if it has one) in the same way it declares its inputs. Then, checking that the max height from the puzzle output matches the bundle's declaration is no different from checking that the bundle's aggregated signature matches the AGG_SIG_ME and AGG_SIG_UNSAFE outputs of the puzzles.

That seems a lot simpler to me, compared to adding another singleton input with who knows what sort of convoluted conditions.

**Comment by bramcohen** [comment]
The same announcement can be leveraged by any number of other coin spends in the same block. The mempool doesn't currently support smushing together two transactions which are spending the same coin in the same way but there's no reason why it couldn't in the future and it's certainly possible for someone at the edge to do it and introduce the combined transaction.

None of the new conditions proposed here help with the auditable randomness which is necessary for scalable lotteries. For that you either need VDF output verification on chain (which can be done already but is ridiculously expensive and hence would need a new opcode to be practical) or a single elimination tournament (which can be done already and hence is probably the right way) or using an n/2 of n threshold signature scheme as a sort of perverse hash function where the signature of the challenge value was used as the random value. Nobody's quite figured out how to do key setup for that one ideally but it's used for some proof of stake systems.

**Comment by bramcohen** [comment]
Some of the princples here are:

The simplicity of the blockchain is much more important than the simplicity of smart coins (which aren't even bad here, because we've got singletons working fine)

Smart coins shouldn't be able to make assertions about the state of other smart coins unless those coins opt into it.

Here's a paper about scalable lotteries: https://arxiv.org/abs/1612.05390

**Comment by SlugFiller** [comment]
@bramcohen 

> The same announcement can be leveraged by any number of other coin spends in the same block.

In theory, but is it any easier on the mempool compared to just checking an extra coin? Also, it still requires a race between all readers, and the one writer. My unspent check suggestion allows setting a "safety buffer" in blocks. A sort of implicit "waiting room" for the writer, in which both the old and new values are valid.

Incidentally, the same issue exists for the "non-spending input" alternative I suggested, unless you allow a non-spending input to co-exist with a spending input, if it is on the same block, although that only gives a 1 block buffer between readers and writer. Alternately, if it's a low-frequency writer, giving a spending input priority, although there's no motivation for the farmer to uphold that.

> it's certainly possible for someone at the edge to do it and introduce the combined transaction.

If not done by the winning farmer, this is just one more spend in the mempool, competing with all the existing ones. It has no more chance to win then one that isn't smushed with anything.

Any other attempts to "trick" a farmer into creating a smushed block only increase unpredictability. For instance, sending the oracle and the oracle-using coin as two spends could result in the other being treated as invalid and thrown out of the mempool, because it asserts a coin that isn't in the spend. Even though a block with both is valid, whether a farmer building a block can detect this is a completely different question. Depending on the set-up, you could be looking at NP-hard block building.

> None of the new conditions proposed here help with the auditable randomness which is necessary for scalable lotteries.

Anything that queries future network data is an "auditable randomness". This is because you can't predict or control it. It would require the cooperation of all farmers and timelords to "force" a certain future network data, and even then, if it's a hash of something, it becomes an NP-hard problem to get the correct hash. It could be as simple as "farmer key at block height" (In theory, that specific metric would tilt the odds towards large farmers, but in practice, if official pools gain the upper hand, no single farmer would be responsible for more than a statistically insignificant fraction of the blocks).

> For that you either need VDF output verification on chain (which can be done already but is ridiculously expensive and hence would need a new opcode to be practical)

The more I think about it, for a "seedless" VDF to work, you need the VDF computation time to be larger than the run time of the lottery. For instance, if the lottery has a 1 week entry time, you need a VDF that would run for 1 week on the best intel processor, and may run 1-3 months on a regular processor. Even setting aside electricity costs to even check if you won, having to wait that long to pick your reward would make the lottery not particularly attractive, even if it does work in theory.

> or a single elimination tournament (which can be done already and hence is probably the right way)

I already explained how this can be manipulated with "strategic forfeiting". You have to set a rule for "What happens if someone doesn't provide their pre-image", and the sane thing to set would be "If they don't give an image within x blocks, they forfeit". The problem is that you don't, and can't, get a strict ordering via pre-images, so a forfeit at an earlier stage of the tournament, can change the winner at a later stage.

> or using an n/2 of n threshold signature scheme as a sort of perverse hash function where the signature of the challenge value was used as the random value. Nobody's quite figured out how to do key setup for that one ideally but it's used for some proof of stake systems.

This one DEFINITELY has "strategic forfeiting". If the challenge is based on who gives their key, then an attacker can opt to give or deny their key to influence the challenge, and hence, the winner. Even if doing so drops you out of the competition, you can still get an increased chance disproportionate to investment via Sybil attack.

> Smart coins shouldn't be able to make assertions about the state of other smart coins unless those coins opt into it.

A non-spending input could achieve that (i.e. A special type of coin input that does not spend the coin, does not allow creating coins, does not add its amount to the output, and requires a special assertion from the puzzle output to indicate this is a non-spending solution). Alternately, an unspent-check that requires a solution, and also requires the checked coin to have a certain output indicating it allows the check. The two are equivalent, other than the possibility of the latter to have a "safety buffer" from an actual spend. Safety input can be artificially achieved  in the prior, by having an "allow non-spending re-spend for x blocks" output to a puzzle, either for the spending puzzle, or for the non-spending puzzle (the prior is probably easier).

Incidentally, since announcements don't have a "to" coin_id (i.e. They're always a block-wide broadcast), and since almost every transaction save for single coin->coin transfer may potentially require announcements to prevent spend bundle manipulation (I actually wanted to make a longer discussion about a potential attack there, but that's a different topic), almost every coin is bound to "opt into" being asserted by other coins sooner or later.

> Here's a paper about scalable lotteries: https://arxiv.org/abs/1612.05390

The way they do it is by having each player release a "chain" of hashes in advance, so that the hash at each stage of the tournament is unknown. This prevents "strategic forfeiting", because you don't know who would win at a later stage, and can't predict the final winner based on an earlier stage. A problem with this is that players need to know the total number of participant before offering their hash chain, so that the hash chain can be the correct length. The reason the original bet needs to be a full chain is because for it to be a lottery, instead of a series of regular bets with increased stakes, it requires that once a person waged anything, they commit to sticking with the tournament, all-or-nothing, until there's a single winner.

It could be worked around by having a "sufficiently long" hash chain, and a matching limit on max participants. e.g. 32 hash chain, with a hard limit of up to 4 billion participants. The solution size of the hash chain can be reduced by simply making it a recursive hash, making each pre-image be the hash of the next pre-image.

**Comment by bramcohen** [comment]
For an n/2 of n threshold signature by definition you only need n/2 of the participants to make the signature but you also need n/2 of them to find out what the signature is so strategic forfeiting would require a truly massive conspiracy. The problem with it as an approach is getting it started without a trusted setup which noone has got working quite right yet.

Having to wait on VDF outputs is inherent. There's some auditable randomness on the blockchain in the form of challenge values because it's already done all the work, although that's only for a relatively short period of time so not the greatest thing, and pulling that information out would be ugly because it doesn't map cleanly to block height. Having a condition which said something about the challenge at some point could go in eventually, but it's in the same category as ones which assert max time: Potentially dangerous and not immediately necessary so they aren't going in for now. The potential for attackers to seriously mess with things which rely on max height by delaying them for long enough that they become permanently unusable is very serious.

**Comment by SlugFiller** [comment]
@bramcohen 

> For an n/2 of n threshold signature by definition you only need n/2 of the participants to make the signature but you also need n/2 of them to find out what the signature is so strategic forfeiting would require a truly massive conspiracy.

No, you just need to wait until n/2-1 have given their signatures, and then quickly pick the final signature, from all your Sybil participants, that gives you the win. It's a bit of a race, but still very far from requiring a massive conspiracy. If you have the processing power, you can also be a bit safer by trying to make 2 picks from the n/2-2 position. So long as signatures don't (and can't) come "all at once", you can always pick the final one.

> Having to wait on VDF outputs is inherent.

Hence the problem. Picking the wait time is crucial. Too short, and someone could do a "last minute wager" attack. Too long, and it would be depressing to wait for the reward. The issue is, rather than there being a soft spot between the two, there's probably some overlap.

> There's some auditable randomness on the blockchain in the form of challenge values because it's already done all the work, although that's only for a relatively short period of time so not the greatest thing

What do you mean? Don't the challenges have to be saved in the blockchain for verification?

> , and pulling that information out would be ugly because it doesn't map cleanly to block height.

This sounds odd to me. Every block in the network has to contain a farmer answer to a challenge. So necessarily, there must be a matching challenge for every block. If not, then I'm misunderstanding something very fundamental about how farming is working.

> The potential for attackers to seriously mess with things which rely on max height by delaying them for long enough that they become permanently unusable is very serious.

1. Delay how? If there's some way to infinitely delay a transaction, that's an issue in and of itself, because you can just as easily delay up to a min height condition as you can to a max height.

2. Max height conditions are actually usable to stop exactly this sort of attack. Have an "infinitely usable" condition, but guard it by a max-height, so if a person tries to live-lock the coin, eventually the max-height is reached, and the coin is "freed" to be used by other conditions.

**Comment by bramcohen** [comment]
Be definition with a unique threshold signature it doesn't matter which participants do the signing, the signature will be the same regardless. BLS has this property, but getting the setup done trustlessly is a bit problematic.

A condition which is usable after a certain point is a min height condition, which we already have.

**Comment by SlugFiller** [comment]
@bramcohen 

> Be definition with a unique threshold signature it doesn't matter which participants do the signing, the signature will be the same regardless. BLS has this property, but getting the setup done trustlessly is a bit problematic.

I'll accept that I don't know enough about BLS signatures to point out the flaws/disadvantages of such a system. Other than a Sybil attack in which the attacker tries to make n/2 of the bets, although that depends on what happens in the event of a "default".

> A condition which is usable after a certain point is a min height condition, which we already have.

That was my point. If you can mess up a contract with unlimited delays, that is a perfect example. In fact, consider the effects on an open lottery coin.

You want people to be able to freely register. Every registration "spends" the coin, creating a new coin with the appended participant list. You have a minimal height in which you expect someone to pick up the winnings, because you don't want them to pick them up whenever their odds are in their favor. For the same reason, you want some space between the last person registering, and the winner picking up the winnings.

Here's the thing, though. If there's no max height on registration, it's possible to live-lock the coin by registering more and more participants, until the minimal height for picking up the reward comes and goes. Then, you can keep adding participants until you have an indication that you're going to win, front-running the actual winner from picking up their reward until that happens.

Even if you add a minimal height condition that "closes registrations", the attack can keep front-running that transaction with more registrations, preventing the registrations from closing until the attacker is ready for them to close.

So rather than having max height creating an attack vector (by live-locking until it is reached), it actually closes an attack vector (by live-locking until the min height is reached).

Now a "height oracle" WOULD work for this situation, IF you can't live-lock the oracle. And that goes back to the question of "How do you read an oracle without spending it?"

**Comment by bramcohen** [comment]
Open registration is inherently problematic. A number of different types of problems could cause difficulties in everyone getting registered in time, which is an argument for letting registrations continue if they haven't petered out yet. In any case it can be done with no max height, all you need is an occasional grace period during which there can't be any more registrations but it can be ended by anyone if the time is at least some end date.

**Comment by SlugFiller** [comment]
@bramcohen 

If the registration period can continue indefinitely, this can be a way to "depress" participants into forfeiting. Lotteries usually have a specific period, like once a week, etc. after which participants expect to know if they've won or not.

There's no merit in "everyone getting registered in time". The whole point of open registration is that there isn't a preset "everyone". Anyone who "misses the chance" essentially lose nothing, nor does an attacker gain anything from blocking registration - the attacker just earns less in the case of a win.

It makes a lot more sense to have a set period, after which, whomever's in is in, whomever's out is out. Those out can always set up their own lottery, if they're that desperate to participate.

**Comment by bramcohen** [comment]
Closing this for being overbroad and undermotivated. More targeted suggestions of individual new conditions with specific behavior and targeted use cases may be considered, but don't expect that to happen quickly.
